// @syncfusion/ej2-base/e2e/protractor.config.js
var shell = require('shelljs');
var os = require('os');
var camelCase = require('pix-diff/lib/camelCase.js');
var seleniumAddress = 'http://150.107.121.2:4444/wd/hub';

var config = require('../../../../config.json');

exports.config = {
    allScriptsTimeout: 600000,

    getPageTimeout: 60000,

    framework: 'jasmine',

    jasmineNodeOpts: {
        defaultTimeoutInterval: 100000
    },
    multiCapabilities: [{
        'browserName': 'chrome',
        'chromeOptions': {
            'args': ['no-sandbox']
        }
    }],

    specs: ['../../../../e2e/tests/**/*.spec.js'],

    onPrepare: function() {
        const PixDiff = require('pix-diff');
        const fs = require('fs');
        const path = require('path');
        browser.ignoreSynchronization = true;
        browser.waitForAngularEnabled = false;
        browser.isDesktop = true;
        browser.basePath = require('../../../../protractor.browser.json').basePath;

        browser.driver.manage().window().setSize(1600, 1200);

        //move to out off user screen
        browser.driver.manage().window().setPosition(2500, 0);

        browser.getCapabilities().then(function(cap) {
            browser.browserName = cap.get('browserName');

            browser.pixResult = PixDiff;

            browser.pixDiff = new PixDiff({
                basePath: './e2e',
                diffPath: './e2e',
                formatImageName: '{tag}'
            });

            //override difference path
            browser.pixDiff.diffPath = path.normalize(camelCase('./e2e/Diff/' + browser.browserName));
            //create folder if not present
            createF(browser.pixDiff.diffPath);

            browser.compareScreen = function(element, fileName, opt) {
                createF(camelCase('e2e/expected/' + browser.pixDiff.browserName));
                createF(camelCase('e2e/actual/' + browser.pixDiff.browserName));

                // thresholdType: 'percent',
                // threshold: 0.009,
                var option = {
                    imageAPath: '/expected/' + browser.pixDiff.browserName + '/' + fileName, // Use file-path 
                    imageBPath: '/actual/' + browser.pixDiff.browserName + '/' + fileName,
                    filter: ['grayScale'],
                    debug: true,
                    hideShift: true,
                };
                var doneFn = arguments[arguments.length - 1];
                if (typeof opt === 'object' && Object.keys(opt).length) {
                    Object.assign(option, opt);
                }
                if (!fs.existsSync(path.resolve(__dirname, '../../../../' + camelCase('e2e/Expected/' + browser.pixDiff.browserName + '/' + fileName) + '.png'))) {
                    browser.pixDiff.saveRegion(element, '/Expected/' + browser.pixDiff.browserName + '/' + fileName, option).then(
                        function() {
                            console.log('Expect Image Created');
                            browser.saveCheckImage(element, fileName, option, doneFn);
                        }
                    );
                } else {
                    browser.saveCheckImage(element, fileName, option, doneFn);
                }

            }
            browser.saveCheckImage = function(element, fileName, option, doneFn) {
                browser.pixDiff.saveRegion(element, '/Actual/' + browser.pixDiff.browserName + '/' + fileName, option).then(() => {
                    browser.pixDiff.checkRegion(element, '/Expected/' + browser.pixDiff.browserName + '/' + fileName, option).then((result) => {
                        //
                        // *  - `RESULT_UNKNOWN`: 0
                        // *  - `RESULT_DIFFERENT`: 1
                        // *  - `RESULT_SIMILAR`: 7
                        // *  - `RESULT_IDENTICAL`: 5
                        expect(result.code).toEqual(browser.pixResult.RESULT_IDENTICAL);
                        if (typeof doneFn === 'function') {
                            doneFn();
                        }
                    });
                });
            }

            browser.waitForEvent = function(id, moduleName, eventName) {
                return browser.executeAsyncScript(function(id, moduleName, eventName, callback) {
                    var instances = document.getElementById(id).ej2_instances;
                    var instanceObj;
                    for (var i = 0; instances && i < instances.length; i++) {
                        if (instances[i].getModuleName() == moduleName) {
                            instanceObj = instances[i]
                        }
                    }
                    if (instanceObj) {
                        var handler;
                        handler = function(e) {
                            instanceObj.removeEventListener(eventName, handler);
                            callback();
                        };
                        instanceObj.addEventListener(eventName, handler);
                    } else {
                        callback();
                    }
                }, id, moduleName, eventName);
            }

            browser.injectScript = function(path) {
                return browser.executeAsyncScript(function(path) {
                    var head = document.getElementsByTagName('head')[0];
                    var script = document.createElement('script');
                    script.type = 'text/javascript';
                    script.onload = function() {
                        callback();
                    }
                    script.src = path;
                    head.appendChild(script);
                }, browser.basePath + path);

            }

            browser.injectCss = function(content) {
                return browser.wait(browser.executeScript(`
                        var style = document.createElement('style');
                        style.id = 'browsercss';
                        if (style.styleSheet) {style.styleSheet.cssText = '` + content + `';}
                        else{style.appendChild(document.createTextNode('` + content + `'));}
                        document.head.appendChild(style);
                        `));
            }

            browser.load = function(path) {
                browser.get(browser.basePath + path);
                if (browser.css) {
                    browser.injectCss(browser.css);
                }
            }
        });

    },
};

//Server Configuration 
if (/jenkins/.test(os.hostname())) {
    exports.config.multiCapabilities.push({
        'browserName': 'internet explorer'
    });
    exports.config.multiCapabilities.push({
        'browserName': 'MicrosoftEdge'
    });
    exports.config.multiCapabilities.push({
        'browserName': 'firefox'
    });
    exports.config.seleniumAddress = seleniumAddress;
} else {
    // Local Testing Purpose
    if (config.browsers && config.browsers.length) {
        for (var i = 0; i < config.browsers.length; i++) {
            var browserName = config.browsers[i];
            exports.config.multiCapabilities.push({
                'browserName': browserName
            });
        }
    } else {
        if (os.platform() === 'win32') {
            exports.config.multiCapabilities.push({
                'browserName': 'internet explorer'
            });
            if (Number(os.release().split('.')[0]) >= 10) {
                exports.config.multiCapabilities.push({
                    'browserName': 'MicrosoftEdge'
                });
            }
        }
        exports.config.multiCapabilities.push({
            'browserName': 'firefox'
        });
    }

    exports.config.seleniumAddress = config.seleniumAddress || 'http://localhost:4444/wd/hub/';
}

function createF(path) {
    shell.mkdir('-p', path);
}
